### ISSUE TITLE
Retrieving results in datalad-run from datalad cmdline calls

### ISSUE THREAD

**Issue by bpoldrack** [datalad/datalad#2996]
We regularly need to call datalad commands via `datalad-run` or `datalad-run-procedure`.
This leads to an independent datalad process and subsequently to the situation, that we don't have access to the results of those secondary datalad commands. For some workflows this is pretty inconvenient. For the things we need in `datalad-hirni` (leading to this issue), we can solve that with a special purpose approach, of course, and there are many ways to do it. However, it seems worth discussing it in a more general fashion and may be come up with a solution in `datalad-run` itself.

Since we leave the python context entirely all those "secondary calls" can do is writing to stdout (or some file, of course). Despite the issue, that `run` currently channels the output through, we could theoretically parse the output (if those commands were called with `-f json`) from within `run` and yield. This could be done with a special output renderer for example. So `-f pipe_json` could mark a line (or a block) in the output, that `run` would then parse into a regular result to yield, while all non-marked output remains untouched. This would require the to be called command or procedure to make sure this renderer is used (which, I think, is fine, since particularly procedures can grow quite complex and it might be necessary for them to decide the results of which commands should be accessible for `run`).
An alternative would be to have some special temporary location (within .git?) to write results to, that `run` would then read. Or having an option for `run` to give a path where to read such result dumps from.
What else seems to be a thinkable approach to you guys?

I lean towards the output marking ATM, since it's easier to get and yield those while it's happening. On the other hand it kind of pollutes the output that is meant for the user especially with procedures which may consist of a mix of datalad and non-datalad calls.

What do you think, @datalad/developers ?

**Comment by mih** [comment]
I think is worthwhile to equip run with this ability. It would become much more practical to use `run` inside datalad commands.

Few aspects that I think are worth keeping in mind when we are moving in this direction.

1. locking or blocking until unlocked: eventually we will want to be able to use datalad in a multi-process fashion. It would make sense to make `run` aware of other `run` calls happening at the same time and allow it to wait from them to finish.
2. non-cmdline "runs": with the growing infrastructure in place within `run`, it is worth thinking about the possibility for non-cmdline runners, ie. Python code (like datalad api calls that run). if we would be able to push regular api calls through run, we might be able to achieve process-level parallelization in a few select places, without breaking a sweat.

On the specific topic here:

I am torn. Adding a line prefix will prevent interoperation with in general-purpose JSON processors that could have a stream of JSON records pipe into them. OTOH: just producing JSON records would force `run` to test convert any output line to see if it could be a datalad result. Maybe that could be combined with a cheap test if a line is a JSON record to speed things up.

My gut seems to suggest that going with `-f json` in combination with a minimum set of criteria that need to be fulfilled to consider a particular JSON record to be a datalad result (that needs to be yielded) could be a viable and still robust approach.

**Comment by bpoldrack** [comment]
> My gut seems to suggest that going with -f json in combination with a minimum set of criteria that need to be fulfilled to consider a particular JSON record to be a datalad result (that needs to be yielded) could be a viable and still robust approach.

Think so, too. WRT to other JSON processors we could have a marker saying "the next X lines are datalad results", leaving the actual lines untouched so they can still be fed into something else, while simultanously `run` can have a comparably cheap test before firing up a JSON parser. Plus: streaming the entire thing into something else that expects a plain JSON stream would still need an in between filter to ignore those marker lines, but this one is simple and could come with datalad as a shell script for example. But note, that this is only relevant if you really need to have it both ways. If it is a dedicated result renderer that is marking those results, you can still use `-f json` as before in a lot of cases.

re 1:
Not sure how to deal with that. Think of a run call to something that calls run itself. Do we want to prohibit that kind of procedures? Or can you think of a way to detect this situation?

re 2:
Agree. Worth mentioning that we do have this concept for `Runner` as well although it's hardly used. Intention there was to be able to apply "protocols" to callables and syscalls.

**Comment by kyleam** [comment]
> However, it seems worth discussing it in a more general fashion and may be come up with a solution in `datalad-run` itself.

Yeah, I think it makes sense to have a general solution for this.

> What else seems to be a thinkable approach to you guys?

I can't think of anything besides the approaches you've already outline.

I too lean towards parsing stdout, and I agree with Michael's gut in terms of starting with `-f json` without markers, optimizing from there as needed.  I'm not sure it's worth the complication, but I also wonder whether we should add a parameter to `run_command` and perhaps `Run.__call__` that controls whether we even consider parsing.

**Comment by bpoldrack** [comment]
I also wonder about another aspect of it. If `run` harvests such a result from stdout, should it then swallow it? Assuming `run` would capture stdout, possibly parse it and then spitting it out again, so actual user output is still going to stdout (from perspective of the `run` caller). If it's meant to be yielded as a result by `run` it doesn't need to go to final output (kind of doubling it), right?

**Comment by mih** [comment]
@christian-monch `run` already uses the `Runner` inside for execution, but in pass-through-mode for stdout/err. Maybe this could be implemented using a dedicated protocol and the generator-style runner that we now have?
